Promise/Promise.swift

```diff
@@ -9,74 +9,89 @@
 import Foundation
 
 struct Callback<Value> {
-    let onFulfilled: (Value -> ())
-    let onRejected: (ErrorType -> ())
-    let queue: dispatch_queue_t
+    let onFulfilled: (Value) -> Void
+    let onRejected: (Error) -> Void
+    let queue: DispatchQueue
     
-    func callFulfill(value: Value) {
-        dispatch_async(queue, {
+    func callFulfill(_ value: Value) {
+        queue.async(execute: {
             self.onFulfilled(value)
         })
     }
     
-    func callReject(error: ErrorType) {
-        dispatch_async(queue, {
+    func callReject(_ error: Error) {
+        queue.async(execute: {
             self.onRejected(error)
         })
     }
 }
 
-enum State<Value>: CustomStringConvertible {
-    case Pending
-    case Fulfilled(value: Value)
-    case Rejected(error: ErrorType)
-    
+enum State<Value> {
+
+    /// The promise has not completed yet.
+    /// Will transition to either the `fulfilled` or `rejected` state.
+    case pending
+
+    /// The promise now has a value.
+    /// Will not transition to any other state.
+    case fulfilled(value: Value)
+
+    /// The promise failed with the included error.
+    /// Will not transition to any other state.
+    case rejected(error: Error)
+
+}
+
+// Convenience methods
+extension State {
+
     var isPending: Bool {
-        if case .Pending = self {
+        if case .pending = self {
             return true
         } else {
             return false
         }
     }
     
     var isFulfilled: Bool {
-        if case .Fulfilled = self {
+        if case .fulfilled = self {
             return true
         } else {
             return false
         }
     }
     
     var isRejected: Bool {
-        if case .Rejected = self {
+        if case .rejected = self {
             return true
         } else {
             return false
         }
     }
     
     var value: Value? {
-        if case let .Fulfilled(value) = self {
+        if case let .fulfilled(value) = self {
             return value
         }
         return nil
     }
     
-    var error: ErrorType? {
-        if case let .Rejected(error) = self {
+    var error: Error? {
+        if case let .rejected(error) = self {
             return error
         }
         return nil
     }
+}
 
-    
+extension State: CustomStringConvertible {
     var description: String {
         switch self {
-        case .Fulfilled(let value):
+        case .fulfilled(let value):
             return "Fulfilled (\(value))"
-        case .Rejected(let error):
+        case .rejected(let error):
             return "Rejected (\(error))"
-        case .Pending:
+        case .pending:
             return "Pending"
         }
     }
```

Comment:
if you don't mind, keep these all in the same declaration instead of each in extensions